### PR TITLE
Add rudimentary support for broadcast via aiosync-client

### PR DESCRIPTION
**Pull request by mzpqnxow**
This is referenced in #291 

I tried to keep this as light as possible, just to be enough to suit my needs (interfacing with a proprietary technology that uses COAP with UDP broadcast as the transport)

If you prefer this be done differently, please let me know. I'm happy to take your guidance. I'm very new to COAP and equally new to the `aiocoap` codebase

Thanks!

### CODE REVIEW

File: aiocoap/protocol.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ async def create_client_context(cls, *, loggername="coap", loop=None, transports
             if transportname == 'udp6':
                 from .transports.udp6 import MessageInterfaceUDP6
                 await self._append_tokenmanaged_messagemanaged_transport(
-                    lambda mman: MessageInterfaceUDP6.create_client_transport_endpoint(mman, log=self.log, loop=loop))
+                    lambda mman: MessageInterfaceUDP6.create_client_transport_endpoint(mman, log=self.log, loop=loop, broadcast=broadcast))
```
Comment (chrysn) [inline aiocoap/protocol.py:162]: Is there any good reason to make this a flag? IOW: Will anything bad come of aiocoap always setting that option?

We're setting a lot of socket options  even though we don't know yet whether we'll need them on this particular occasion. (For example, V6ONLY=0 is set even though we may not use IPv4 at all).

File: aiocoap/protocol.py [hunk 2]
```diff
@@ -241,13 +241,11 @@ async def create_server_context(cls, site, bind=None, *, loggername="coap-server
                     lambda mman: MessageInterfaceSimple6.create_client_transport_endpoint(mman, log=self.log, loop=loop))
             elif transportname == 'tinydtls':
                 from .transports.tinydtls import MessageInterfaceTinyDTLS
-
```
Comment (chrysn) [inline aiocoap/protocol.py:244]: Please don't change whitespace or other unrelated code in PRs. If this is your editor's doing, you may want to reign it in on code bases that don't follow that editor's particular coding style.

File: aiocoap/tokenmanager.py [hunk 3]
```diff
@@ -256,7 +256,13 @@ def request(self, request):
         self.outgoing_requests[key] = request
         request.on_interest_end(functools.partial(self.outgoing_requests.pop, key, None))
 
-'''
+'''Not implemented
```
Comment (chrysn) [inline aiocoap/tokenmanager.py:259]: See above on coding style.

File: aiocoap/transports/udp6.py [hunk 4]
```diff
@@ -247,11 +247,20 @@ def _local_port(self):
         return self.transport.get_extra_info('socket').getsockname()[1]
 
     @classmethod
-    async def _create_transport_endpoint(cls, sock, ctx: interfaces.MessageManager, log, loop, multicast=[]):
+    async def _create_transport_endpoint(cls, sock, ctx: interfaces.MessageManager, log, loop, multicast=[], broadcast: bool = False):
         try:
             sock.setsockopt(socket.IPPROTO_IPV6, socknumbers.IPV6_RECVPKTINFO, 1)
         except NameError:
             raise RuntimeError("RFC3542 PKTINFO flags are unavailable, unable to create a udp6 transport.")
+        if broadcast is True:
```
Comment (chrysn) [inline aiocoap/transports/udp6.py:255]: See above: Can this be unconditional?

File: aiocoap/transports/udp6.py [hunk 5]
```diff
@@ -247,11 +247,20 @@ def _local_port(self):
...
+            try:
+                # This shouldn't be needed, at least not when using --broadcast
+                # with aiocoap-client.
+                if not sock.getsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST):
```
Comment (chrysn) [inline aiocoap/transports/udp6.py:259]: Checking and then setting when not yet set is no faster or better than just setting it.

File: aiocoap/transports/udp6.py [hunk 6]
```diff
@@ -247,11 +247,20 @@ def _local_port(self):
...
+                    sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+            except Exception:
+                log.fatal("Unable to set socket to SO_BROADCAST; setsockopt() failed")
+                raise
```
Comment (chrysn) [inline aiocoap/transports/udp6.py:263]:
```suggestion
                log.warning("Unable to set socket to SO_BROADCAST; continuing, but requests can not be sent to broadcast addresses.")
```

When there is no extra flag, this can best-effort.

File: aiocoap/transports/udp6.py [hunk 7]
```diff
@@ -299,9 +308,15 @@ async def _create_transport_endpoint(cls, sock, ctx: interfaces.MessageManager,
         return protocol
 
     @classmethod
-    async def create_client_transport_endpoint(cls, ctx: interfaces.MessageManager, log, loop):
+    async def create_client_transport_endpoint(cls, ctx: interfaces.MessageManager, log, loop, broadcast: bool = False):
         sock = socket.socket(family=socket.AF_INET6, type=socket.SOCK_DGRAM)
         sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
+        if broadcast is True:
```
Comment (chrysn) [inline aiocoap/transports/udp6.py:314]: This is redundant with the code in _create_transport_endpoint, which is called right after this.

File: aiocoap/protocol.py (hunk 1 above)
Comment (mzpqnxow) [reply aiocoap/protocol.py:162]: Very good question. I *think* the downside is that superuser privileges (or possibly some CAP_* on Linux) are required to set this particular ioctl, but I need to check on that

File: aiocoap/protocol.py (hunk 2 above)
Comment (mzpqnxow) [reply aiocoap/protocol.py:244]: Yep, sorry 'bout that, this one (or two) slipped through. I managed to avoid my urge to run black against the entire codebase and submit that as a PR 😜

**Comment by chrysn** [comment]
> best-effort/don't raise on failure

We have to pick one of these because we can't be sure whether a v4 address is broadcast or not, right? If so, best-effort setting it is probably an OK way. I'd like to redo the transport configuration at some point so that there's a configurable way to do this that doesn't need to thread flags through many layers, but I'm not fully clear there yet. (Could be a TOML-ish data structure that tells which protocols to bind to which ports, would allow binding to multiple ports, and setting things like multicast groups, and setting flags such as SO_REUSEPORT or this one).

File: aiocoap/protocol.py (hunk 1 above)
Comment (mzpqnxow) [reply aiocoap/protocol.py:162]: I think I'm misremembering regarding the requirement of additional privileges, the EPERM was coming from an earlier version of this patch when I was using SO_BINDTODEVICE

You may wonder why on earth that would be necessary- the server software I was dealing with actually _required_ the universal broadcast address 255.255.255.255 for a specific endpoint, and I'm working in a multi-homed environment with multiple interfaces having +BROADCAST

In the end I decided it would be outside the scope of this PR to add an option to force the interface, especially without any discussion. My final conclusion was to not attempt to solve that problem with changes to aiocoap; there are a few workarounds good enough for that exceptional case

tl; dr; your suggestion on defaulting to SO_BROADCAST seems to be a good one as far as I can tell

**Comment by chrysn** [comment]
On Tue, May 16, 2023 at 05:32:44AM -0700, AG wrote:
> You may wonder why on earth that would be necessary- the server
> software I was dealing with actually _required_ the universal
> broadcast address 255.255.255.255 for a specific endpoint, and I'm
> working in a multi-homed environment with multiple interfaces having
> +BROADCAST

Given that we're on the udp6 transport that works in IPv6 semantics with
all its benefits, do you really need to bind the interface, or can't you
use the tuple ('::ffff:255.255.255.255', 5684, 0, interface_id) to send
the request out on the right interface? That should even work with URI
notation as `coap://[::ffff:255.255.255.255%eth0]`.

File: aiocoap/protocol.py (hunk 2 above)
Comment (mzpqnxow) [reply aiocoap/protocol.py:244]: > The concern is duly noted, and as soon as I have the number of large branches down a bit, and have verified that no code is becoming less readable, black *will* be the one color in which you get aiocoap, and CI will mark any specks of color with a big red cross mark.

The world wants to know- when will we have PEP-484 across the codebase and no red marks next to mypy? 😉

On a serious note though, I'd be happy to help with that effort- for the public interfaces or the entire project- *if* you're interested obviously, and if time permits

Personally I now find it nearly impossible to write good interfaces without it, though that may be evidence of my lack of discipline rather than evidence of the broad value of type annotations
